lib/redmine/scm/adapters/xitolite_adapter.rb

```diff
@@ -0,0 +1,572 @@
+            @tags = io.readlines.sort!.map { |t| t.strip }
+          end
+          @tags
+        rescue ScmCommandAborted => e
+          logger.error(e.message)
+          []
+        end
+
+
+        def default_branch
+          bras = self.branches
+          return nil if bras.nil?
+          default_bras = bras.select { |x| x.is_default == true }
+          return default_bras.first.to_s if !default_bras.empty?
+          master_bras = bras.select { |x| x.to_s == 'master' }
+          master_bras.empty? ? bras.first.to_s : 'master'
+        end
+
+
+        def entry(path = nil, identifier = nil)
+          parts = path.to_s.split(%r{[\/\\]}).select { |n| !n.blank? }
+          search_path = parts[0..-2].join('/')
+          search_name = parts[-1]
+          if search_path.blank? && search_name.blank?
+            # Root entry
+            Entry.new(path: '', kind: 'dir')
+          else
+            # Search for the entry in the parent directory
+            es = entries(search_path, identifier, report_last_commit: false)
+            es ? es.detect { |e| e.name == search_name } : nil
+          end
+        end
+
+
+        def entries(path = nil, identifier = nil, options = {})
+          path ||= ''
+          p = scm_iconv(@path_encoding, 'UTF-8', path)
+          entries = Entries.new
+          cmd_args = %w|ls-tree -l|
+          cmd_args << "HEAD:#{p}"          if identifier.nil?
+          cmd_args << "#{identifier}:#{p}" if identifier
+          git_cmd(cmd_args) do |io|
+            io.each_line do |line|
+              e = line.chomp.to_s
+              if e =~ /^\d+\s+(\w+)\s+([0-9a-f]{40})\s+([0-9-]+)\t(.+)$/
+                type = $1
+                sha  = $2
+                size = $3
+                name = $4
+                if name.respond_to?(:force_encoding)
+                  name.force_encoding(@path_encoding)
+                end
+                full_path = p.empty? ? name : "#{p}/#{name}"
+                n      = scm_iconv('UTF-8', @path_encoding, name)
+                full_p = scm_iconv('UTF-8', @path_encoding, full_path)
+                entries << Entry.new({:name => n,
+                 :path => full_p,
+                 :kind => (type == "tree") ? 'dir' : 'file',
+                 :size => (type == "tree") ? nil : size,
+                 :lastrev => options[:report_last_commit] ?
+                                 lastrev(full_path, identifier) : Revision.new
+                }) unless entries.detect { |entry| entry.name == name }
+              end
+            end
+          end
+          entries.sort_by_name
+        rescue ScmCommandAborted => e
+          logger.error(e.message)
+          []
+        end
+
+
+        def lastrev(path, rev)
+          return nil if path.nil?
+          cmd_args = %w|log --no-color --encoding=UTF-8 --date=iso --pretty=fuller --no-merges -n 1|
+          cmd_args << rev if rev
+          cmd_args << "--" << path unless path.empty?
```

Comment:
Prefer single-quoted strings when you don't need string interpolation or special symbols.